File: algorithms/algorithms-python/leetcode/Question_0025_Reverse_Nodes_in_k_Group.py
#!/usr/bin/python
# coding: utf-8

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        pre, next, last, i = dummy, None, None, 0
        while head:
            i += 1
            next = head.next
            if i % k == 0:
                head.next = None
                last = pre.next
                pre.next = self.reverse(pre.next)
                last.next = next
                pre = last
            head = next
        return dummy.next


    def reverse(self, head):
        pre, next = None, None
        while head:
            next = head.next
            head.next = pre
            pre = head
            head = next
        return pre

File: algorithms/algorithms-cplusplus/leetcode/Question_0025_Reverse_Nodes_in_k_Group.cc

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
 public:
  ListNode* reverseKGroup(ListNode* head, int k) {
    if (!head || k <= 1) {
      return head;
    }
    ListNode dummy(0);
    dummy.next = head;
    int count = 0;
    ListNode *pre = &dummy, *cur = head;
    while (cur) {
      count++;
      if (count % k == 0) {
        pre = reverse(pre, cur->next);
        cur = pre->next;
      } else {
        cur = cur->next;
      }
    }
    return dummy.next;
  }

  ListNode* reverse(ListNode* pre, ListNode* next) {
    ListNode *last = pre->next, *cur = pre->next->next;
    while (cur != next) {
      last->next = cur->next;
      cur->next = pre->next;
      pre->next = cur;
      cur = last->next;
    }
    return last;
  }
};
